README
Win32::SerialPort and Win32API::CommPort
VERSION=0.20, 1 April 2010
VERSION=0.21, 30 April 2010 (Makefile.PL and test improvements)

Hello Serial Port users:

If you are not running Windows, you want the Device::SerialPort module
instead of this one. It has a compatible user interface and runs on
Operating Systems that fully support POSIX.pm. Available from your
favorite CPAN site. Since someone asked, MS-DOS does NOT support POSIX.pm.
But OSX does.

It has been over 10 years since the last release. No functions are
considered "experimental" anymore. And this release is mostly test and
installer improvements along with a few bug fixes. There is now a module
Test::Device::SerialPort on CPAN that emulates these modules (and their
POSIX cousins) for testing application code without hardware. While the
modules themselves have supported USB ports, COM10 and above, VirtualBox,
and embedded ports, the tests did not up to now.

These modules are intended for Win32 ports of Perl without requiring
a compiler or using XS. In every case, compatibility has been selected
over performance. Since everything is (convoluted but still pure) perl,
you can fix flaws and change limits if required. But please file a bug
report if you do. I have tested with the ActiveState (5.6.x, 5.8.x, 5.10.x)
and Strawberry (5.8.x and 5.10.x) distributions. But these modules should
work on any Win32 version that supports Aldo Calpini's Win32::API.
While the modules have worked on earlier perl versions, I no longer have
any way to test them for compatibility.

This is the sixth production release. All planned features are now
implemented.  If you see any place where the code does not match the
documentation, consider it a bug and please report it.

FILES:

    Changes		- for history lovers
    Makefile.PL		- the "starting point" for traditional reasons
    MANIFEST		- file list
    README.txt		- this file (CRLF)
    README    		- same file with CPAN-friendly name (LF only)
    eg/any_os.plx	- cross-platform use and init
    eg/demo1.plx	- talks to a "really dumb" terminal
    eg/demo2.plx	- "poor man's" readline and chat
    eg/demo3.plx	- looks like a setup menu - but only looks :-(
    eg/demo4.plx	- simplest setup: "new", "required param", "restart"
    eg/demo5.plx	- "waitfor" and "nextline" using lookfor
    eg/demo6.plx	- basic tied FileHandle operations, record separators
    eg/demo7.plx	- a Perl/Tk based terminal, event loop and callbacks
    eg/demo8.plx	- command line terminal emulator with Term::Readkey
    eg/demo9.plx	- using debug on a close()
    eg/options.plx	- post-install test that prints available options
    eg/stty.plx		- first try at Unix lookalike

    lib				- install directory
    lib/Win32			- install directory
    lib/Win32/SerialPort.pm	- the reason you're reading this
    lib/Win32API		- install directory
    lib/Win32API/CommPort.pm	- the raw API calls and other internals

    html			- install directory
    html/Win32			- install directory
    html/Win32/SerialPort.html	- documentation
    html/Win32API		- install directory
    html/Win32API/CommPort.html	- documentation

    t			- test directory
    t/Altport.pm	- stub for inheritance test
    t/test1.t		- RUN ME FIRST, tests and creates configuration
    t/test2.t		- tests restarting_a_configuration and timeouts
    t/test3.t		- Inheritance and export version of test1.t
    t/test4.t		- Inheritance version of test2.t and "restart"
    t/test5.t		- tests to optional exports from CommPort
    t/test6.t		- stty tests
    t/test7.t		- tied FileHandle tests

This is a genuine production release. You can complain if I don't
maintain compatibility from this point forward. The experimental
items might change a bit - but I'll try to avoid breaking any code.

PRE-INSTALL and TEST:

Run 'perl Makefile.PL' first with nothing connected to "COM1". You can
specify a diferent port to test with 'perl Makefile.PL TESTPORT=PORT'.
I recommend you use a "traditional" hard-wired port for testing if
available. Several tests are skipped for USB and other specialized ports.

For those with make (e.g. Strawberry Perl), the normal mantra applies:
	perl Makefile.PL TESTPORT=port
	make
	make test
	make install

Makefile.PL also generates two scripts for those who don't have a version
of 'make' (e.g. nmake, dmake):
	perl nomake_test
	perl nomake_install

Test::More and related routines are used to generate reports. The test
suite covers most of the module methods and leaves the port set for 9600
baud, 1 stop, 8 data, no parity, no handshaking, and other defaults. At
various points in the testing, it expects unconnected CTS and DTR lines.
The final configuration is saved by test1.t as port_test.cfg in the
current directory (e.g COM1_test.cfg).

The pod2html output is stored in the html directory. Some perl versions
support automatically installing html and others do not.

Tests may also be run individually by typing:
	'perl test?.t [COMx]'

All tests are expected to pass - I would be very interested in hearing
about failures ("not ok"). These tests should be run from a command
line (DOS box).

DEMO PROGRAMS:

Connect a dumb terminal (or a PC that acts like one) to COM1 and setup
the equivalent configuration. Starting demo1.plx should print a three
line message on both the terminal and the Win32 command line. The
terminal keyboard (only) now accepts characters which it prints to both
screens until a CONTROL-Z is typed. Also included is demo2.plx - a truly
minimal chat program. Bi-directional communication without an event loop,
sockets, pipes (or much utility ;-) This one uses CAPITAL-Q from the
active keyboard to quit since <STDIN> doesn't like CONTROL-Z. And each
command shell acts a little differently (Cygnus "bash", COMMAND.COM).
Try running the terminal at 4800 baud to get errors (or 300 to get
"breaks").

AltPort.pm and test3.t implement the "basic Inheritance test" discussed
in perltoot and other documentation. It also imports the :STAT constants.
It's otherwise only slightly modified from test1.t (you'll get a different
"alias" if you run test2.t or demo3.plx after test3.t). There are some
subtle functional changes between test2.t and test4.t. But test4.t also
calls CommPort methods directly rather than through SerialPort and adds
tests for lookfor and stty_xxx methods.

You can read (many of the important) settings with demo3.plx. If you
give it a (valid) configuration file on the command line, it will open
the port with those parameters (and "initialized" set - so you can test
simple changes: see the parity example at the end of demo3.plx).

Run options.plx to see the available choices for various parameters
along with the current values. If you have trouble, I will probably
ask you to save the output of options.plx in a file and send it to me.
You can specify a port name for options.plx on the command line
(e.g. 'perl options.plx COM2').

Demo4.plx is a "minimum" script showing just the basics needed to get
started.

Demo5.plx demonstrates various uses of the lookfor routine including
setups for "waitfor" and a primitive "readline". Try them out. The
default "stty" settings work with a VT-100 style terminal. You may
have to set the options by hand. Use any editor. Let me know if the
descriptions in the documentation are useable. And if any more options
are necessary.

Demo6.plx demonstrates tied FileHandles. Perl 5.005 is recommended.
It "requires" 5.004. It implements timeouts on all user inputs - so
you can run it "hands-off" to see what happens.

Demo7.plx uses Tk to create a terminal emulator. Its included to show
polling and callbacks using an event loop.

Demo8.plx is a simple command-line terminal emulator contributed by
Andrej Mikus.

Stty.plx is a wrapper around the stty method that implements a clone
of the Unix/POSIX function of the same name. It's line noise unless
you know Unix.

The Perl Journal #13 included an article on Controlling a Modem with
Win32::SerialPort. Perl Journal #16 showed how to work with the X10
series of home automation modules (ControlX10::).

Please tell me what does and what doesn't work. The module has proven
to be pretty robust. But I can't test all possible configurations.
Don't trust it for anything important without complete testing.

Available on CPAN under Author BBIRTH or on gitHub:

	http://github.com/wbirthisel/win32-serialport

Thanks,

-bill

Copyright (C) 2010, Bill Birthisel. All rights reserved. This module is
free software; you can redistribute it and/or modify it under the same
terms as Perl itself.


Changes
Revision history for Perl extension Win32API::SerialPort.

0.01  Tue Apr 14 12:01:47 1998
	- original version; created by h2xs 1.18
	- command: h2xs -X -n SerialPort

0.02  Thu Apr 16 15:53:30 1998
	- extensive edits to pod section
	- consolidate pre-existing code fragments
	- post documentation for comments

0.03  Sun Apr 26 22:24:47 1998
	- add test1.pl
	- extensive bug fixes while developing tests

0.04  Mon May 25 15:14:30 1998
	- significant additions to test1.pl
	- add test2.pl, demo1.pl, demo2.pl, README.txt
	- first version with read and write functions
	- first public alpha

0.05  Sun May 31 21:26:02 1998
	- rewrite tests using "is_ok", "is_bad", "is_zero" subroutines
	- abort "start" on parameter validation failure
	- rename tests to ".t" and demos to ".plx"
	- add fourth return value to status and reset_error.
	- Add built-in error messages and use in demo?.plx
	- add read_bg, write_bg, read_done, write_done (background I/O)

0.06  Fri Jun 05 17:38:20 1998
	- make sure baudrate and databits interpreted as integers
	- Add AltPort.pm and test3.t for Inheritance testing
	- Add demo3.plx as setup example
	- Clone README.txt to preferred CPAN name
	- Add Makefile.PL (as stub) for CPAN
	- Add parity_enable. Other DCB bits not implemented yet

0.07  Tue Jun 16 22:43:20 1998
	- Add test4.t for Inheritance testing
	- Makefile.PL runs Test::Harness on 5.004 (GS)
	- Export :COMM constants on request
	- Fixes for status on NT (differs from Win95) - add BM_AllBits
	- Purge now clears read/write busy
	- Save error_msg/user_msg in configuration file

0.08  Wed Jun 17 21:56:11 1998
	- Add demo4.plx
	- Add description of "required" parameters
	- Fix bug that "required" all the timing and buffer parameters
	- Small fixes to test?.t for "-w" warnings

0.09  Fri Aug 21 xx:xx:xx 1998
	- First pre-beta, no general distribution
	- Rename :COMM exports to :STAT
	- Make "handshake" recommended rather than required
	- Add restart
	- Add demo5.plx
	- Split into Win32:SerialPort(user) and Win32API::CommPort(raw)

0.10  Sat Aug 29 23:18:50 1998
	- First public beta
	- Add demo6.plx and test5.t
	- Add :RAW :COMPROP :PARAM :DCB :ALL exports to CommPort.pm
	- Add $ResetEvent to CommPort.pm
	- Fix a number of timing bugs

0.11  Mon Sep 14 23:40:00 1998
	- Add current baud to ok list
	- add error messages
	- allow set_write_buf(0)
	- limited release to test TPJ examples and bug fixes

0.12  Sun Nov 08 23:40:00 1998
	- MAJOR UPDATE - Second public beta
	- Modify CommPort.pm to export subroutine wrappers instead of
	  API objects under tag "RAW:". Many related changes in test5.t
	- Add perl version test to Makefile.PL
	- Add Install.PL and blib subdirectory tree (5.004 and above)
	- Add options.plx to aid in debugging after installation
	- Update MANIFEST, README and documentation
	- Add error messages for failed "save" and "write_settings" 
	- Fix bugs in range checking for "baudrate" and "buffers"
	- Rename API objects to "$_object"
	- Delete $_GetLocalError
	- Use "my $dcb" instead of $self->{"_DCB"} in CommPort.pm
	- Move test suite files to "t" subirectory (5.004 and above)
	- Add "use blib" to the test suite
	- Revise demo*.plx to use proper Namespaces
	- Create "Build_3xx" subdirectory for non-Makamaker test suite
	- Add manual install instructions for ActiveState build 3xx

0.13  Sat Nov 28 19:35:05 1998
	- Third public beta
	- Add lookfor, lookclear, lastlook, are_match, input-loopback,
	  is_prompt, and stty-like functions (MAJOR ENHANCEMENT)
	- Consolidate demo5.plx and demo6.plx using lookfor et. al.
	- Rewrite Makefile.PL and Install.PL to eliminate need for
	  "Build_3xx" subdirectory, add test suite support for Build_3xx,
	  and "use lib" instead of "use blib" (which CPAN reserves for
	  binary distributions)
	- Add html install for 5.004 and above
	- All t/test* files modified - test4.plx extensively (testing
	  new functions)

0.14  Sun Feb 07 23:41:00 1999
	- Fourth public beta
	- Add Tied FileHandle support (MAJOR ENHANCEMENT)
	- Several enhancments to lookfor:
	    Explicit Regular Expression support much like expect
	    Block until $count characters received option
            Both match_pattern and input_that_matched returned by lastlook
	- Fix bug in CloseHandle with prevented reopening port in same
	  process. Add return value to "close" method.
	- Revise Makefile.PL to improve test suite support for Build_3xx
	- Add "$no_silly_warnings" to CommPort for Build_3xx
	- Numerous documentation changes
	- Delete never-used "dosmode" constructor
	- add demo6.plx and t/test6.t for tied FileHandles

0.15  Sun May 09 16:25:40 1999
	- First production release.
	- Add $RBUF_Size and internal_buffer().
	- Add $quiet option to new.
	- Correct bugs in handshake off control masks.
	- Rename "eof" and "kill" to s_eof and s_kill to reduce confusion with
	  Perl builtin functions.
	- Add stty method and opost parameter.
	- Default stty_xxx setup for raw port rather than dumb terminal.
	- Add compiled regex support for are_match.
	- Add streamline, lastline, linesize, matchclear.
	- Add $instead to lastlook.
	- Recode lookfor to improve efficiency.
	- Improve error handling on write timeout.
	- Add cntl_char and argv_char subs to support stty.
	- Add test suite port option to Makefile.PL
	- Extensive test suite additions
	- small changes in demo4.plx amd demo6.plx, large ones in demo5.plx
	- new demo7.plx, demo8.plx, stty.plx
	- on CPAN as Version 0.150 due to upload problem

0.16  Sun July 18 15:17:15 1999
	- Second production release.
	- Add "pulse_xxx" functions to support CM17 device on MisterHouse
	- Remove "list context" support from binary methods in CommPort.
	  Most were read_only anyway. Reflected in test1.t and test3.t.
	- Disable stty_eof unless stty_icanon (bug).
	- Add methods for the seldom used DCB bits like ignore_null.

Makefile.PL
use 5.006;
use strict;
use warnings;
use ExtUtils::MakeMaker;

# named Win32:: for a good reason. Recommendation from FAQ
die "No support for OS\n" unless ($^O eq 'MSWin32');

my $version = "0.21";	# for nomake versions

my @SERIALS;

# Scan for argument we know about
my @passargs;
if (@ARGV) {
	foreach my $arg (@ARGV) {
		my ($key,$value)=split /=/, $arg, 2;
		if ($key eq "TESTPORT") {
			push(@SERIALS,$value);
		}
		else {
			push(@passargs,$arg);
		}
	}
}
@ARGV=@passargs;

my $file = "COM1";
if (@SERIALS) {
    # only first one is relevant
    $file = shift @SERIALS;
}
# confirm a test device seems to exist (presumes passed Win32 test above)
# tested with cmd and bash shells
if (system('echo', 'test', '>', $file)) {
	warn "Device $file does not appear to exist. Testing impossible\n";
	exit (0);
} else {
	warn "found result=$?, file=$file\n";
}

print <<COMMENTS;
            Win32::SerialPort and Win32API::CommPort
                        VERSION $version

A 'Makefile' created for those with 'make' or CPAN.pm users.
It will test using PORT = $file. To test using a different PORT,
run again specifying: 'perl Makefile.PL TESTPORT=<PORT>'
The normal 'Mantra' would then apply:
        make
        make test
        make install

For those without 'make' or an equivalent like 'nmake' or 'dmake' there
are perl-only scripts which do the same things:
        Test with:    perl nomake_test
        Install with: perl nomake_install

Test with nothing connected to $file.
PORT is not verified present and accessible until tests run.
Timeout tests can take up to 30 seconds per test.
COMMENTS

my $dfile = "t/DefaultPort.pm";
unlink $dfile;
print "\nCreating new $dfile\n";
open my $default, '>', $dfile or die "Can't create $dfile: $!\n";
print {$default} <<EOF;
package SerialJunk;
# Created by Makefile.PL

\$Makefile_Test_Port = "$file";
EOF

close $default;


$dfile = "nomake_test";
unlink $dfile;
print "Creating new $dfile\n";
open $default, '>', $dfile or die "Can't create $dfile: $!\n";

print {$default} <<"TEST4";
# Created by Makefile.PL
# VERSION $version

use Test::Harness;
runtests ("t/test1.t", "t/test2.t", "t/test3.t", "t/test4.t",
	  "t/test5.t", "t/test6.t", "t/test7.t");

# To run individual tests, type:
#     C:\\> perl t/test1.t [ $file ]
# Continue with 'perl t/test2.t' through 'perl t/test7.t'
# See README and other documentation for additional information.
TEST4

close $default;

$dfile = "nomake_install";
unlink $dfile;
print "Creating new $dfile\n";
open $default, '>', $dfile or die "Can't create $dfile: $!\n";

print {$default} <<"INST4";
# Created by Makefile.PL
# VERSION $version

use 5.006;

use Config;
use strict;
use warnings;
use ExtUtils::Install qw( install );

my \$FULLEXT = "Win32/SerialPort";
my \$INST_LIB = "./lib";
my \$HTML_LIB = "./html";

my \$html_dest = "";	# edit real html base here if autodetect fails

if (exists \$Config{installhtmldir} ) {
    \$html_dest = "\$Config{installhtmldir}";
}
elsif (exists \$Config{installprivlib} ) {
    \$html_dest = "\$Config{installprivlib}";
    \$html_dest =~ s%\\\\lib%\\\\html%;
}

if ( length (\$html_dest) ) {
    \$html_dest .= '\\lib\\site';
    install({
	   read => "\$Config{sitearchexp}/auto/\$FULLEXT/.packlist",
	   write => "\$Config{installsitearch}/auto/\$FULLEXT/.packlist",
	   \$INST_LIB => "\$Config{installsitelib}",
	   \$HTML_LIB => "\$html_dest"
	  },1,0,0);
}
else {
    warn "Can't find html base directory. Skipping doc install\\n";
    install({
	   read => "\$Config{sitearchexp}/auto/\$FULLEXT/.packlist",
	   write => "\$Config{installsitearch}/auto/\$FULLEXT/.packlist",
	   \$INST_LIB => "\$Config{installsitelib}",
	  },1,0,0);
}

__END__
INST4

close $default;
my %mm_args=(
    'NAME'          => 'Win32::SerialPort',
    'VERSION_FROM' => 'lib/Win32/SerialPort.pm', # finds $VERSION
    'PREREQ_PM'     =>	{ 'Test::More' => 0,
			  'Win32::API' => 0,
			  'Win32' => 0
			}, # e.g., Module::Name => 1.1
    'ABSTRACT_FROM' => 'lib/Win32/SerialPort.pm', # retrieve abstract from module
    'AUTHOR'        => 'Bill Birthisel <wcbirthisel@alum.mit.edu>',
    'LIBS'          => [''], # e.g., '-lm'
    'DEFINE'        => '', # e.g., '-DHAVE_SOMETHING'
    'INC'           => '-I.', # e.g., '-I. -I/usr/include/other'
    'clean'         => { 'FILES' => "t/DefaultPort.pm *.cfg t/*.cfg" },
    'realclean'     => { 'FILES' => "nomake_test nomake_install" },
);

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(%mm_args);
